File: src/main/java/tmi/TooManyItems.java
package tmi;

import arc.Core;
import arc.Events;
import arc.util.Time;
import mindustry.Vars;
import mindustry.content.Items;
import mindustry.game.EventType;
import mindustry.mod.Mod;
import tmi.recipe.RecipeItemManager;
import tmi.recipe.RecipeType;
import tmi.recipe.RecipesManager;
import tmi.recipe.parser.*;
import tmi.ui.*;
import tmi.util.Consts;
import tmi.util.KeyBinds;

public class TooManyItems extends Mod {
  public static RecipesManager recipesManager;
  public static RecipeItemManager itemsManager;
  public static ModAPI api;

  public static RecipesDialog recipesDialog;
  public static SchematicDesignerDialog schematicDesigner;
  public static BatchBalanceDialog batchBalance;
  public static KeyBinds binds;

  public TooManyItems() {
    api = new ModAPI();
    recipesManager = new RecipesManager();
    itemsManager = new RecipeItemManager();

    ConsumerParser.registerVanillaConsumeParser();
    registerRecipeParser();

    Events.on(EventType.ClientLoadEvent.class, e -> Time.runTask(0, () -> {
      EntryAssigner.assign();
      Vars.ui.settings.game.checkPref("tmi_button", true);
      Vars.ui.settings.graphics.sliderPref("tmi_gridSize", 150, 50, 300, 10, Integer::toString);

      api.afterInit();
    }));
  }

  private void registerRecipeParser() {
    //几乎所有的原版游戏工厂方块的分析工具
    recipesManager.registerParser(new GenericCrafterParser());
    recipesManager.registerParser(new UnitFactoryParser());
    recipesManager.registerParser(new ReconstructorParser());
    recipesManager.registerParser(new UnitAssemblerParser());
    recipesManager.registerParser(new PumpParser());
    recipesManager.registerParser(new SolidPumpParser());
    recipesManager.registerParser(new FrackerParser());
    recipesManager.registerParser(new DrillParser());
    recipesManager.registerParser(new BeamDrillParser());
    recipesManager.registerParser(new SeparatorParser());
    recipesManager.registerParser(new GeneratorParser());
    recipesManager.registerParser(new ConsGeneratorParser());
    recipesManager.registerParser(new HeatGeneratorParser());
    recipesManager.registerParser(new ThermalGeneratorParser());
    recipesManager.registerParser(new VariableReactorParser());
    recipesManager.registerParser(new HeatCrafterParser());
    recipesManager.registerParser(new HeatProducerParser());
    recipesManager.registerParser(new AttributeCrafterParser());
    recipesManager.registerParser(new WallCrafterParser());
  }

  @Override
  public void init() {
    Consts.load();
    Cursor.init();
    RecipeType.init();

    binds = new KeyBinds();
    recipesDialog = new RecipesDialog();
    //TODO: not usable yet
    if (Core.settings.getBool("tmi_enable_preview")) schematicDesigner = new SchematicDesignerDialog();
    batchBalance = new BatchBalanceDialog();

    binds.load();
    api.init();

    recipesManager.init();
  }
}

File: src/main/kotlin/tmi/TooManyItems.kt
package tmi

import arc.Events
import arc.files.Fi
import arc.util.Time
import mindustry.Vars
import mindustry.entities.bullet.LaserBulletType
import mindustry.game.EventType.ClientLoadEvent
import mindustry.mod.Mod
import mindustry.ui.dialogs.BaseDialog
import mindustry.world.blocks.defense.turrets.LaserTurret
import tmi.recipe.RecipeItemManager
import tmi.recipe.RecipesManager
import tmi.recipe.parser.*
import tmi.ui.Cursor
import tmi.ui.EntryAssigner
import tmi.ui.TmiUI
import tmi.util.KeyBinds
import universe.ui.markdown.Markdown
import universe.ui.markdown.MarkdownStyles

class TooManyItems : Mod() {
  companion object {
    val modFile: Fi by lazy {
      Vars.mods.getMod(TooManyItems::class.java)!!.root
    }

    @JvmField
    var recipesManager: RecipesManager = RecipesManager()
    @JvmField
    var itemsManager: RecipeItemManager = RecipeItemManager()
    @JvmField
    var api: ModAPI = ModAPI()
    @JvmField
    val binds = KeyBinds()
  }

  init {
    ConsumerParser.registerVanillaConsumeParser()
    registerDefaultParser()

    Events.on(ClientLoadEvent::class.java) {
      Time.runTask(0f) {
        EntryAssigner.assign()
        Vars.ui.settings.game.checkPref("tmi_button", true)
        Vars.ui.settings.game.checkPref("tmi_items_pane", false)
        Vars.ui.settings.graphics.sliderPref("tmi_gridSize", 150, 50, 300, 10) { i -> i.toString() }
        api.afterInit()

        sample()
      }
    }
  }

  private fun sample(){
    val dialog = BaseDialog("markdown")
    val markdown = Markdown(
$$"""
      一个辅助mod，为您提供快速检索材料，单位，建筑和矿$物等的生产与采$集引导信息，让您不再因为安装了太多mod而不知所措（缩写为 `$`）
      """.trimIndent(),
      MarkdownStyles.defaultMD
    )

    dialog.addCloseButton()
    dialog.cont.add(markdown).grow().pad(20f)
    dialog.show()
  }

  private fun registerDefaultParser() {
    //Parser for the vanilla game factory blocks
    recipesManager.registerParser(GenericCrafterParser())
    recipesManager.registerParser(UnitFactoryParser())
    recipesManager.registerParser(ReconstructorParser())
    recipesManager.registerParser(UnitAssemblerParser())
    recipesManager.registerParser(ConstructorParser())
    recipesManager.registerParser(PumpParser())
    recipesManager.registerParser(SolidPumpParser())
    recipesManager.registerParser(FrackerParser())
    recipesManager.registerParser(DrillParser())
    recipesManager.registerParser(BeamDrillParser())
    recipesManager.registerParser(SeparatorParser())
    recipesManager.registerParser(GeneratorParser())
    recipesManager.registerParser(ConsumeGeneratorParser())
    recipesManager.registerParser(ImpactReactorParser())
    recipesManager.registerParser(HeatGeneratorParser())
    recipesManager.registerParser(ThermalGeneratorParser())
    recipesManager.registerParser(VariableReactorParser())
    recipesManager.registerParser(HeatCrafterParser())
    recipesManager.registerParser(HeatProducerParser())
    recipesManager.registerParser(AttributeCrafterParser())
    recipesManager.registerParser(WallCrafterParser())
    recipesManager.registerParser(ItemTurretParser())
    recipesManager.registerParser(LiquidTurretParser())
    recipesManager.registerParser(PowerTurretParser())
    recipesManager.registerParser(BuildingParser())
  }

  override fun init() {
    Cursor.init()
    binds.load()
    api.init()

    recipesManager.init()

    TmiUI.init()
  }
}
